pkg/cmd/pulumi/pulumi.go

```diff
@@ -393,14 +393,28 @@ func checkForUpdate(ctx context.Context) *diag.Diag {
 		return nil
 	}
 
-	latestVer, oldestAllowedVer, err := getCLIVersionInfo(ctx)
-	if err != nil {
-		logging.V(3).Infof("error fetching latest version information "+
-			"(set `%s=true` to skip update checks): %s", env.SkipUpdateCheck.Var().Name(), err)
+	var skipUpdateCheck bool
+	latestVer, oldestAllowedVer, err := getCachedVersionInfo()
+	if err == nil {
+		// If we have cached version information then don't bother warning, we would have warned when we last
+		// updated the cache. But we will still log this to the internal logging system that by default users
+		// don't see.
+		skipUpdateCheck = true
+	} else {
```

Comment:
nit: This could be an early exit and simplify the branching here and in the remainder of the function which is dependent on mutable state.

Reply:
+1 to early return here.
Long-term, it appears that with this change the cache file is useless as a source of information, and is just a marker. We can (here or in a future change) stop storing the JSON-encoded blob in it, and instead just use its last modified time to determine whether we've already informed the user.

Reply:
> +1 to early return here.

Can't early return because I still wanted to do the version check and write to the logging log.

> Long-term, it appears that with this change the cache file is useless as a source of information, and is just a marker.

Yes, I did consider changing that all with this change but wasn't sure if anything else might use this file.